### PR TITLE
documenting minimum requirements for implementing spark api's

### PR DESCRIPTION
**Pull request by GoEddie**
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [-] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

Hi,

I thought it would be helpful to start documenting what is required to implement new functions, I think this is right from the pr's that I have done so far.

ed

### CODE REVIEW

File: docs/coding-guidelines/new-functions.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+* Function name is correct
+* Number of parameters are correct
+* Parameters are of the correct type
+* Function return is of the correct type
+
+The typical approach is to use `Assert.IsType<>` and call a function:
+
+ 
+    ```C#
```
Comment (suhsteve) [inline docs/coding-guidelines/new-functions.md]: I think the indenting of the ```s are off.  The markdown isn't rendered correctly

Comment (GoEddie) [reply docs/coding-guidelines/new-functions.md]: i've removed the s p a c e s :)

File: docs/coding-guidelines/new-functions.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+Minium requirements for implementing Spark API functions
```
Comment (rapoth) [inline docs/coding-guidelines/new-functions.md]: nit: 
```suggestion
Minimum Requirements for Implementing Spark API Functions
```

File: docs/coding-guidelines/new-functions.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+======================================================
+
+This document describes the minimum requirements for contributing new Spark API functions to dotnet/spark.
```
Comment (rapoth) [inline docs/coding-guidelines/new-functions.md]:
```suggestion
This document describes the minimum requirements for contributing new Spark API functions to [dotnet/spark](https://github.com/dotnet/spark).
```

File: docs/coding-guidelines/new-functions.md (hunk 2 above)
Comment (rapoth) [inline docs/coding-guidelines/new-functions.md]: Many thanks for writing this down! :)

File: docs/coding-guidelines/new-functions.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+Comments
+--------
+
+Each Spark API has a comment that is used to generate the documentation on the [.NET API Browser](https://docs.microsoft.com/en-gb/dotnet/api/?view=spark-dotnet). The recommended approach is to start with the comments that are used by the Apache Spark comments for each function and then expand when required.
```
Comment (rapoth) [inline docs/coding-guidelines/new-functions.md]:
```suggestion
Each Spark API has a comment that is used to generate the documentation on the [.NET API Browser](https://docs.microsoft.com/en-gb/dotnet/api/?view=spark-dotnet). The recommended approach is to start with comments that are used by the respective API you are attempting to expose from the Apache Spark code and then improvise when required. For instance, see the comments for the original implementation of `agg` in [Scala](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala#L223) and compare against [C#](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/RelationalGroupedDataset.cs#L37).
```

File: docs/coding-guidelines/new-functions.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+To find the original Apache Spark function's comment you can either go directly to the source code such as [Column Apache Spark implementation](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Column.scala) or you can view the [Apache Spark API documentation](https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.Column).
```
Comment (rapoth) [inline docs/coding-guidelines/new-functions.md]:
```suggestion
To find the original Apache Spark's comment for an API, you can either go directly to the source code such as [Column.scala](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Column.scala) in Apache Spark Core or you can view the respective [Column documentation](https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.Column) on the Apache Spark website.
```

File: docs/coding-guidelines/new-functions.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+Unit Tests
+----------
+
+The approach for unit testing is to check whether the functions have been implemented correctly in dotnet and not to verify that the actual Apache Spark implementation is correct. This approach means that each function should have a unit test that verifies that the:
```
Comment (rapoth) [inline docs/coding-guidelines/new-functions.md]:
```suggestion
The approach for unit testing is to check whether the functions have been implemented correctly in .NET and not to verify that the actual Apache Spark implementation is correct (since this would have been tested in Apache Spark Core already). This approach means that each function should have a unit test that verifies that the following is satisfied:
```

File: docs/coding-guidelines/new-functions.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+* Function return is of the correct type
+
+The typical approach is to use `Assert.IsType<>` and call a function:
+
+
+```C#
+Assert.IsType<Column>(Col("col2"));
+```
+
+For these unit tests see [the IPC Functions Tests](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/FunctionsTests.cs#L41).
```
Comment (rapoth) [inline docs/coding-guidelines/new-functions.md]:
```suggestion
For these unit tests, see the [IPC Functions Tests](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/FunctionsTests.cs#L41).
```

File: docs/contributing.md [hunk 8]
```diff
@@ -7,3 +7,8 @@ Coding Style
 We intend to bring dotnet/spark into full conformance with the following style guidelines:
    * [C# Coding Style](coding-guidelines/csharp-coding-style.md)
    * [Scala Coding Style](coding-guidelines/scala-coding-style.md)
+
+Implementing Spark API functions
+--------------------------------
+When you implement one of the Spark API functions that is currently not implemented you should include comments and unit tests, refer to the [Minium requirements for implementing Spark API functions](coding-guidelines/new-functions.md)
```
Comment (rapoth) [inline docs/contributing.md]:
```suggestion
When you implement one of the Spark API functions that is currently not implemented, you should include comments and unit tests, refer to the [Minimum Requirements for Implementing Spark API Functions](coding-guidelines/new-functions.md) mini-guide.
```

File: docs/coding-guidelines/new-functions.md (hunk 2 above)
Comment (GoEddie) [reply docs/coding-guidelines/new-functions.md]: no problem, I just thought it might help other people :)
